File: src/main/java/com/handwoong/everyonewaiter/config/aop/LoggingAop.java
package com.handwoong.everyonewaiter.config.aop;

import java.lang.reflect.Method;
import lombok.extern.slf4j.Slf4j;
import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.AfterReturning;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.aspectj.lang.annotation.Pointcut;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.stereotype.Component;
import org.springframework.transaction.support.TransactionSynchronizationManager;

@Slf4j
@Aspect
@Component
public class LoggingAop {

    @Pointcut("execution(* com.handwoong.everyonewaiter.service.*.*(..))")
    private void servicePointcut() {
    }

    @Before("servicePointcut()")
    public void parameterLogBefore(JoinPoint joinPoint) {
        Method method = getMethod(joinPoint);
        log.info("============== Method Start : {} ==============", method.getName());

        Object[] args = joinPoint.getArgs();
        for (Object arg : args) {
            log.info("Parameter : '{}' '{}'", arg.getClass().getSimpleName(), arg);
        }
    }

    @AfterReturning(value = "servicePointcut()", returning = "returnValue")
    public void afterReturning(JoinPoint joinPoint, Object returnValue) {
        Method method = getMethod(joinPoint);
        log.info("============== Method End : {} ==============", method.getName());
        if (returnValue != null) {
            log.info("Return : '{}' '{}'", returnValue.getClass().getSimpleName(), returnValue);
        }
    }

    @Around("servicePointcut()")
    public Object serviceLog(ProceedingJoinPoint joinPoint) throws Throwable {
        String transactionName = TransactionSynchronizationManager.getCurrentTransactionName();
        long start = System.currentTimeMillis();
        log.info("============ [{}] Transaction Start ============", transactionName);
        Object result = joinPoint.proceed();
        long end = System.currentTimeMillis();
        log.info("========= [{}] Transaction End '{}'ms  =========", transactionName, end - start);
        return result;
    }

    private Method getMethod(JoinPoint joinPoint) {
        MethodSignature signature = (MethodSignature) joinPoint.getSignature();
        return signature.getMethod();
    }
}

File: src/main/kotlin/com/handwoong/everyonewaiter/config/aop/LoggingAop.kt
package com.handwoong.everyonewaiter.config.aop

import com.handwoong.everyonewaiter.util.logger
import org.aspectj.lang.JoinPoint
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.*
import org.aspectj.lang.reflect.MethodSignature
import org.springframework.stereotype.Component
import org.springframework.transaction.support.TransactionSynchronizationManager
import java.lang.reflect.Method

@Aspect
@Component
class LoggingAop {

    private val log = logger()

    @Pointcut("execution(* com.handwoong.everyonewaiter.service.*.*(..)) && !@annotation(com.handwoong.everyonewaiter.util.ExcludeLog)")
    private fun servicePointcut() {
    }

    @Before("servicePointcut()")
    fun parameterLogBefore(joinPoint: JoinPoint) {
        val method = getMethod(joinPoint)
        log.info("============== Method Start : ${method.name} ==============")
        for (arg in joinPoint.args) {
            log.info("Parameter : '${arg.javaClass.simpleName}' '${arg}'")
        }
    }

    @AfterReturning(value = "servicePointcut()", returning = "returnValue")
    fun afterReturning(joinPoint: JoinPoint, returnValue: Any?) {
        val method = getMethod(joinPoint)
        log.info("Return : '${returnValue?.javaClass?.simpleName}' '${returnValue}'")
        log.info("============== Method End : ${method.name} ==============")
    }

    @Around("servicePointcut()")
    fun serviceLog(joinPoint: ProceedingJoinPoint): Any? {
        val transactionName = TransactionSynchronizationManager.getCurrentTransactionName()
        val startTime = System.currentTimeMillis()

        log.info("============ [${transactionName}] Transaction Start ============")
        val result = joinPoint.proceed()
        val endTime = System.currentTimeMillis()
        log.info("========= [${transactionName}] Transaction End '${endTime - startTime}'ms  =========")

        return result
    }

    private fun getMethod(joinPoint: JoinPoint): Method {
        val methodSignature = joinPoint.signature as MethodSignature
        return methodSignature.method
    }

}
